aot/src/main/java/com/dylibso/chicory/aot/AotCompiler.java

```diff
@@ -636,272 +597,73 @@ private void compileFunction(
-        ctx.enterScope(FUNCTION_SCOPE, FunctionType.of(List.of(), type.returns()));
+        // track targets to detect backward jumps
+        Set<Long> visitedTargets = new HashSet<>();
 
         // compile the function body
-        int exitBlockDepth = -1;
-        for (int idx = 0; idx < body.instructions().size(); idx++) {
-            var ins = body.instructions().get(idx);
-
-            Label label = labels.get(idx);
-            if (label != null) {
-                asm.visitLabel(label);
-            }
-
-            // skip instructions after unconditional control transfer
-            if (exitBlockDepth >= 0) {
-                if (ins.depth() > exitBlockDepth) {
-                    continue;
-                }
-                if (ins.opcode() != OpCode.ELSE && ins.opcode() != OpCode.END) {
-                    continue;
-                }
-
-                exitBlockDepth = -1;
-                if (ins.opcode() == OpCode.END) {
-                    ctx.scopeRestoreStackSize();
-                }
-            }
-
+        for (AotInstruction ins : instructions) {
             switch (ins.opcode()) {
-                case NOP:
-                    break;
-                case BLOCK:
-                case LOOP:
-                    ctx.enterScope(ins.scope(), blockType(ins));
-                    break;
-                case END:
-                    ctx.exitScope(ins.scope());
-                    break;
-                case UNREACHABLE:
-                    exitBlockDepth = ins.depth();
-                    emitInvokeStatic(asm, THROW_TRAP_EXCEPTION);
-                    asm.visitInsn(Opcodes.ATHROW);
-                    break;
-                case RETURN:
-                    exitBlockDepth = ins.depth();
-                    emitReturn(asm, type, ctx);
-                    break;
-                case IF:
-                    ctx.popStackSize();
-                    ctx.enterScope(ins.scope(), blockType(ins));
-                    asm.visitJumpInsn(Opcodes.IFEQ, labels.get(ins.labelFalse()));
-                    // use the same starting stack sizes for both sides of the branch
-                    if (body.instructions().get(ins.labelFalse() - 1).opcode() == OpCode.ELSE) {
-                        ctx.pushStackSizesStack();
+                case LABEL:
+                    Label label = labels.get(ins.operand(0));
+                    if (label != null) {
+                        asm.visitLabel(label);
+                        visitedTargets.add(ins.operand(0));
                     }
                     break;
-                case ELSE:
-                    asm.visitJumpInsn(Opcodes.GOTO, labels.get(ins.labelTrue()));
-                    ctx.popStackSizesStack();
-                    break;
-                case BR:
-                    exitBlockDepth = ins.depth();
-                    if (ins.labelTrue() < idx) {
+                case GOTO:
+                    if (visitedTargets.contains(ins.operand(0))) {
```

Comment:
Why are we skipping the interruption check here?
Can we document what's the rationale?

Reply:
There's a comment above that I hoped would explain it
```java
// track targets to detect backward jumps
Set<Long> visitedTargets = new HashSet<>();
```
If the target label has already been visited, then this is a backward jump. Otherwise, it is a forward jump. We document this in `InterpreterMachine.checkinterruption()`:

> Terminate WASM execution if requested. This is called at the start of each call and at any potentially backwards branches. Forward branches and other non-branch instructions are not checked, as the execution will run until it eventually reaches a termination point.

Should we reference this comment in `AotMethods.checkInterruption()`?

Reply:
> Should we reference this comment in `AotMethods.checkInterruption()`?

yes please :pray: